pkg/client/unversioned/portforward/portforward.go

```diff
@@ -279,34 +284,42 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 	headers.Set(api.StreamType, api.StreamTypeData)
 	dataStream, err := pf.streamConn.CreateStream(headers)
 	if err != nil {
-		glog.Errorf("Error creating forwarding stream for port %d -> %d: %v", port.Local, port.Remote, err)
+		util.HandleError(fmt.Errorf("error creating forwarding stream for port %d -> %d: %v", port.Local, port.Remote, err))
 		return
 	}
-	// Send a Reset when this function exits to completely tear down the stream here
-	// and in the remote server.
-	defer dataStream.Reset()
+	defer dataStream.Close()
 
 	go func() {
-		// Copy from the remote side to the local port. We won't get an EOF from
-		// the server as it has no way of knowing when to close the stream.  We'll
-		// take care of closing both ends of the stream with the call to
-		// stream.Reset() when this function exits.
-		if _, err := io.Copy(conn, dataStream); err != nil && err != io.EOF && !strings.Contains(err.Error(), "use of closed network connection") {
-			glog.Errorf("Error copying from remote stream to local connection: %v", err)
+		// Copy from the remote side to the local port.
+		if _, err := io.Copy(conn, dataStream); err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
+			util.HandleError(fmt.Errorf("error copying from remote stream to local connection: %v", err))
 		}
+
+		// inform the select below that the remote copy is done
 		doneChan <- struct{}{}
```

Comment:
To cover what we discussed on irc: with a change to wait for both copies to complete before returning, when running `ab -n 5 -c 1000 ...` where `n` is even a number as small as 5, the local -> remote copies seem to "hang". They don't unblock, and `ab` times out. If I revert to what is in this PR, I can get `n` up to 200 or a bit more. Any larger, and there are issues using the Docker socket to inspect the container (which happens for each locally forwarded connection).

In summary, I think we need to short-circuit and return if there is an error copying local -> remote, but otherwise wait for the remote -> local copy to finish before returning.